ccl/crosscluster/physical: TestStandbyReadTSPollerJob failed

ccl/crosscluster/physical.TestStandbyReadTSPollerJob [failed](https://mesolite.cluster.engflow.com/invocations/default/54ae3aad-9890-411d-be15-8902df428ecf?testReportRun=1&testReportShard=13&testReportAttempt=1#targets-Ly9wa2cvY2NsL2Nyb3NzY2x1c3Rlci9waHlzaWNhbDpwaHlzaWNhbF90ZXN0) on master @ [9497f05a5c1b316fbc5fe6cd833ed43e6c118a79](https://github.com/cockroachdb/cockroach/commits/9497f05a5c1b316fbc5fe6cd833ed43e6c118a79):


```
=== RUN   TestStandbyReadTSPollerJob
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestStandbyReadTSPollerJob2930340887
    test_log_scope.go:81: use -show-logs to present logs inline
    standby_read_ts_poller_job_test.go:62: error scanning '&{0xc00cddfe60 <nil>}': pq: database "defaultdb" does not exist
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestStandbyReadTSPollerJob2930340887
--- FAIL: TestStandbyReadTSPollerJob (6.20s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=13</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStandbyReadTSPollerJob.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-42634

ccl/crosscluster/physical.TestStandbyReadTSPollerJob [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17104605?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17104605?buildTab=artifacts#/) on master @ [9497f05a5c1b316fbc5fe6cd833ed43e6c118a79](https://github.com/cockroachdb/cockroach/commits/9497f05a5c1b316fbc5fe6cd833ed43e6c118a79):


```
=== RUN   TestStandbyReadTSPollerJob
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/4efabb294d4906fb514344b5cc79411a/logTestStandbyReadTSPollerJob3764899432
    test_log_scope.go:81: use -show-logs to present logs inline
    standby_read_ts_poller_job_test.go:62: error scanning '&{<nil> 0x4002a43cb0}': sql: no rows in result set
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/4efabb294d4906fb514344b5cc79411a/logTestStandbyReadTSPollerJob3764899432
--- FAIL: TestStandbyReadTSPollerJob (8.34s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStandbyReadTSPollerJob.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

ccl/crosscluster/physical.TestStandbyReadTSPollerJob [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17105662?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17105662?buildTab=artifacts#/) on master @ [9497f05a5c1b316fbc5fe6cd833ed43e6c118a79](https://github.com/cockroachdb/cockroach/commits/9497f05a5c1b316fbc5fe6cd833ed43e6c118a79):


```
=== RUN   TestStandbyReadTSPollerJob
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/4efabb294d4906fb514344b5cc79411a/logTestStandbyReadTSPollerJob1923384503
    test_log_scope.go:81: use -show-logs to present logs inline
    standby_read_ts_poller_job_test.go:62: error scanning '&{0x400d76f680 <nil>}': pq: database "defaultdb" does not exist
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/4efabb294d4906fb514344b5cc79411a/logTestStandbyReadTSPollerJob1923384503
--- FAIL: TestStandbyReadTSPollerJob (4.13s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStandbyReadTSPollerJob.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

ccl/crosscluster/physical.TestStandbyReadTSPollerJob [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17108121?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17108121?buildTab=artifacts#/) on master @ [1de4cbdad151df45e665c86e494d0432db2f8fdf](https://github.com/cockroachdb/cockroach/commits/1de4cbdad151df45e665c86e494d0432db2f8fdf):


```
=== RUN   TestStandbyReadTSPollerJob
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/4efabb294d4906fb514344b5cc79411a/logTestStandbyReadTSPollerJob2085860838
    test_log_scope.go:81: use -show-logs to present logs inline
    standby_read_ts_poller_job_test.go:62: error scanning '&{0x40116db8c0 <nil>}': pq: database "defaultdb" does not exist
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/4efabb294d4906fb514344b5cc79411a/logTestStandbyReadTSPollerJob2085860838
--- FAIL: TestStandbyReadTSPollerJob (2.50s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStandbyReadTSPollerJob.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

ccl/crosscluster/physical.TestStandbyReadTSPollerJob [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17108320?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17108320?buildTab=artifacts#/) on master @ [1de4cbdad151df45e665c86e494d0432db2f8fdf](https://github.com/cockroachdb/cockroach/commits/1de4cbdad151df45e665c86e494d0432db2f8fdf):


```
=== RUN   TestStandbyReadTSPollerJob
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/4efabb294d4906fb514344b5cc79411a/logTestStandbyReadTSPollerJob3258672922
    test_log_scope.go:81: use -show-logs to present logs inline
    standby_read_ts_poller_job_test.go:62: error scanning '&{<nil> 0x40063d2510}': sql: no rows in result set
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/4efabb294d4906fb514344b5cc79411a/logTestStandbyReadTSPollerJob3258672922
--- FAIL: TestStandbyReadTSPollerJob (6.75s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStandbyReadTSPollerJob.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

ccl/crosscluster/physical.TestStandbyReadTSPollerJob [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17109444?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17109444?buildTab=artifacts#/) on master @ [1084e341ee8406f8de908fbaa93d82032f9696c0](https://github.com/cockroachdb/cockroach/commits/1084e341ee8406f8de908fbaa93d82032f9696c0):


```
=== RUN   TestStandbyReadTSPollerJob
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/4efabb294d4906fb514344b5cc79411a/logTestStandbyReadTSPollerJob1501419603
    test_log_scope.go:81: use -show-logs to present logs inline
    standby_read_ts_poller_job_test.go:62: error scanning '&{0x400daec900 <nil>}': pq: database "defaultdb" does not exist
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/4efabb294d4906fb514344b5cc79411a/logTestStandbyReadTSPollerJob1501419603
--- FAIL: TestStandbyReadTSPollerJob (8.00s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStandbyReadTSPollerJob.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/ccl/crosscluster/physical/standby_read_ts_poller_job.go
+++ b/pkg/ccl/crosscluster/physical/standby_read_ts_poller_job.go
@@ -68,6 +68,7 @@ func (r *standbyReadTSPollerResumer) poll(ctx context.Context, execCfg *sql.Exec
 		case <-ctx.Done():
 			return ctx.Err()
 		case <-ticker.C:
+			ticker.Reset(standbyReadTSPollInterval.Get(&execCfg.Settings.SV))
 			tenantInfoAccessor := execCfg.UpgradeJobDeps.SystemDeps().TenantInfoAccessor
 			if tenantInfoAccessor == nil {
 				return errors.AssertionFailedf("tenant info accessor cannot be nil")

--- a/pkg/ccl/crosscluster/physical/standby_read_ts_poller_job_test.go
+++ b/pkg/ccl/crosscluster/physical/standby_read_ts_poller_job_test.go
@@ -10,7 +10,6 @@ import (
 	"fmt"
 	"net/url"
 	"testing"
-	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/ccl/crosscluster/replicationtestutils"
@@ -27,6 +26,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/errors"
 	"github.com/stretchr/testify/require"
 )
@@ -35,8 +35,7 @@ func TestStandbyReadTSPollerJob(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	skip.WithIssue(t, 131611)
-
+	beginTS := timeutil.Now()
 	ctx := context.Background()
 	args := replicationtestutils.DefaultTenantStreamingClustersArgs
 	args.EnableReaderTenant = true
@@ -47,6 +46,7 @@ func TestStandbyReadTSPollerJob(t *testing.T) {
 
 	jobutils.WaitForJobToRun(c.T, c.SrcSysSQL, jobspb.JobID(producerJobID))
 	jobutils.WaitForJobToRun(c.T, c.DestSysSQL, jobspb.JobID(ingestionJobID))
+	t.Logf("test setup took %s", timeutil.Since(beginTS))
 
 	srcTime := c.SrcCluster.Server(0).Clock().Now()
 	c.WaitUntilReplicatedTime(srcTime, jobspb.JobID(ingestionJobID))
@@ -58,25 +58,17 @@ func TestStandbyReadTSPollerJob(t *testing.T) {
 	readerTenantName := fmt.Sprintf("%s-readonly", args.DestTenantName)
 	c.ConnectToReaderTenant(ctx, readerTenantID, readerTenantName, 0)
 
-	pollInterval := time.Microsecond
-	serverutils.SetClusterSetting(t, c.DestCluster, "bulkio.stream_ingestion.standby_read_ts_poll_interval", pollInterval)
-
-	var jobID jobspb.JobID
-	c.ReaderTenantSQL.QueryRow(t, `
-SELECT job_id 
-FROM crdb_internal.jobs 
-WHERE job_type = 'STANDBY READ TS POLLER'
-`).Scan(&jobID)
-	jobutils.WaitForJobToRun(t, c.ReaderTenantSQL, jobID)
-
 	c.SrcTenantSQL.Exec(t, `
 USE defaultdb;
 CREATE TABLE a (i INT PRIMARY KEY);
 INSERT INTO a VALUES (1);
 `)
+	waitForPollerJobToStart(t, c, ingestionJobID)
 
-	// Verify that updates have been replicated to reader tenant
-	testutils.SucceedsWithin(t, func() error {
+	now := timeutil.Now()
+	// Verify that updates have been replicated to reader tenant. This may take a
+	// second as the historical timestamp these AOST queries run needs to advance.
+	testutils.SucceedsSoon(t, func() error {
 		var numTables int
 		c.ReaderTenantSQL.QueryRow(t, `SELECT count(*) FROM [SHOW TABLES]`).Scan(&numTables)
 
@@ -91,7 +83,40 @@ INSERT INTO a VALUES (1);
 				1, actualQueryResult)
 		}
 		return nil
-	}, time.Minute)
+	})
+	t.Logf("waited for %s for updates to reflect in reader tenant", timeutil.Since(now))
+}
+
+func waitForPollerJobToStart(
+	t *testing.T, c *replicationtestutils.TenantStreamingClusters, ingestionJobID int,
+) {
+	// TODO(annezhu): we really should be waiting for the AOST timestamp on the
+	// reader tenant to advance to point where the reader tenant's defaultdb
+	// exists, but we don't have a way to track that. The historical ts really
+	// should be stored in the reader tenant job progress.
+	srcTime := c.SrcCluster.Server(0).Clock().Now()
+	c.WaitUntilReplicatedTime(srcTime, jobspb.JobID(ingestionJobID))
+
+	testutils.SucceedsSoon(t, func() error {
+		var numJobs int
+		c.ReaderTenantSQL.QueryRow(t, `
+SELECT count(*)
+FROM crdb_internal.jobs 
+WHERE job_type = 'STANDBY READ TS POLLER';
+`).Scan(&numJobs)
+
+		if numJobs != 1 {
+			return errors.Errorf("expected 1 standby read ts poller job, but got %d instead", numJobs)
+		}
+		return nil
+	})
+	var jobID jobspb.JobID
+	c.ReaderTenantSQL.QueryRow(t, `
+SELECT job_id
+FROM crdb_internal.jobs 
+WHERE job_type = 'STANDBY READ TS POLLER'
+`).Scan(&jobID)
+	jobutils.WaitForJobToRun(t, c.ReaderTenantSQL, jobID)
 }
 
 func TestFastFailbackWithReaderTenant(t *testing.T) {

--- a/pkg/ccl/crosscluster/replicationtestutils/testutils.go
+++ b/pkg/ccl/crosscluster/replicationtestutils/testutils.go
@@ -636,6 +636,8 @@ var defaultSrcClusterSetting = map[string]string{
 	// Large timeout makes test to not fail with unexpected timeout failures.
 	`stream_replication.stream_liveness_track_frequency`: `'2s'`,
 	`stream_replication.min_checkpoint_frequency`:        `'1s'`,
+	// Finer grain checkpoints to keep replicated time close to present.
+	`physical_replication.producer.timestamp_granularity`: `'100ms'`,
 	// Make all AddSSTable operation to trigger AddSSTable events.
 	`kv.bulk_io_write.small_write_size`: `'1'`,
 	`jobs.registry.interval.adopt`:      `'1s'`,

--- a/pkg/ccl/crosscluster/replicationtestutils/testutils.go
+++ b/pkg/ccl/crosscluster/replicationtestutils/testutils.go
@@ -208,6 +208,14 @@ func (c *TenantStreamingClusters) ConnectToReaderTenant(
 		TenantName:  roachpb.TenantName(tenantName),
 		UseDatabase: "defaultdb",
 	})
+	c.DestSysSQL.Exec(c.T, `ALTER TENANT $1 SET CLUSTER SETTING jobs.registry.interval.adopt = '1s'`, tenantName)
+	c.DestSysSQL.Exec(c.T, `ALTER TENANT $1 SET CLUSTER SETTING bulkio.stream_ingestion.standby_read_ts_poll_interval = '500ms'`, tenantName)
+
+	// Attempt to keep the reader tenant's historical aost close to the present.
+	c.DestSysSQL.Exec(c.T, `ALTER TENANT $1 SET CLUSTER SETTING kv.closed_timestamp.target_duration = '100ms'`, tenantName)
+	c.DestSysSQL.Exec(c.T, `ALTER TENANT $1 SET CLUSTER SETTING kv.rangefeed.closed_timestamp_refresh_interval = '200ms'`, tenantName)
+	c.DestSysSQL.Exec(c.T, `ALTER TENANT $1 SET CLUSTER SETTING kv.closed_timestamp.side_transport_interval = '50ms'`, tenantName)
+
 	require.NoError(c.T, err)
 	c.ReaderTenantConn = c.DestCluster.Server(server).SystemLayer().SQLConn(c.T,
 		serverutils.DBName("cluster:"+tenantName+"/defaultdb"))
@@ -554,6 +562,7 @@ func WaitUntilStartTimeReached(t *testing.T, db *sqlutils.SQLRunner, ingestionJo
 func WaitUntilReplicatedTime(
 	t *testing.T, targetTime hlc.Timestamp, db *sqlutils.SQLRunner, ingestionJobID jobspb.JobID,
 ) {
+	now := timeutil.Now()
 	testutils.SucceedsSoon(t, func() error {
 		err := requireReplicatedTime(targetTime, jobutils.GetJobProgress(t, db, ingestionJobID))
 		if err == nil {
@@ -571,6 +580,7 @@ func WaitUntilReplicatedTime(
 		}
 		return err
 	})
+	t.Logf("waited for %s to advance to %s", timeutil.Since(now), targetTime)
 }
 
 func requireReplicatedTime(targetTime hlc.Timestamp, progress *jobspb.Progress) error {

